### PR TITLE
dpdk: Eliminate unused metadata fields

### PR DESCRIPTION
**Pull request by kamleshbhalui**
The PSA and PNA architecture require a lot of architecture-specific meta-data fields, but most of them are not actually used by a typical P4 program.

If an architecture-specific meta-data field is defined in the P4 because it is required by the architecture (e.g. PSA or PNA), but this field is not used at all by the P4 program, then this field should not be carried into the output .spec file generated by the P4C-DPDK compiler.

### CODE REVIEW

File: backends/dpdk/dpdkAsmOpt.h [hunk 1]
```diff
@@ -135,6 +135,29 @@ class RemoveLabelAfterLabel : public Transform {
     }
 };
 
+
+// This pass Collects all metadata struct member used in program
+class CollectUsedMetadataField : public Inspector {
+    ordered_set<cstring> used_fields;
+  public:
+    bool preorder(const IR::Member *m) override {
+        if (m->toString().substr(0, 2) == "m.") {
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.h:144]: You could use startsWith() 
https://github.com/p4lang/p4c/blob/main/lib/cstring.h#L206

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 2]
```diff
@@ -163,4 +163,32 @@ const IR::IndexedVector<IR::DpdkAsmStatement> *RemoveLabelAfterLabel::removeLabe
...
+const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
+    auto usedFieldsCollector = new CollectUsedMetadataField();
+    usedFieldsCollector->setCalledBy(this);
+    p->apply(*usedFieldsCollector);
+    auto used_fields = usedFieldsCollector->getUsedFields();
+
+    IR::IndexedVector<IR::DpdkStructType> usedStruct;
+    for (auto st : p->structType) {
+        IR::IndexedVector<IR::StructField> usedMetadataFields;
+        for (auto field : st->fields) {
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.cpp:176]: We should use some early exit condition to avoid looping on header structs, Probably, annotations can be checked. The annotation "__metadata__" is added to the metadata struct and since there is a single metadata struct at this point, this loop will be taken only once.

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 3]
```diff
@@ -163,4 +163,32 @@ const IR::IndexedVector<IR::DpdkAsmStatement> *RemoveLabelAfterLabel::removeLabe
...
+            if (used_fields.count("m." + field->name.name)) {
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.cpp:177]: While storing the fields names in used_fields, "m." can be truncated. So this check would become simpler.

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 4]
```diff
@@ -163,4 +163,32 @@ const IR::IndexedVector<IR::DpdkAsmStatement> *RemoveLabelAfterLabel::removeLabe
...
+                usedMetadataFields.push_back(field);
+            }
+        }
+        if (!usedMetadataFields.empty()) {
+            auto newSt = new IR::DpdkStructType(st->srcInfo, st->name,
+                                             st->annotations, usedMetadataFields);
+            usedStruct.push_back(newSt);
+        }
+        else {
```
Comment (usha1830) [inline backends/dpdk/dpdkAsmOpt.cpp:186]: If none of the metadata fields are used, do you intend to return the original struct? We should be creating an empty struct in that.

**Review by usha1830 — CHANGES_REQUESTED** [review]
@kamleshbhalui  Looks fine overall. Have a few minor comments. Please check.

**Comment by usha1830** [comment]
Note: I have not reviewed all the spec files. Briefly looked at first few.

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 2 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.cpp]: done

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 3 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.cpp]: done

File: backends/dpdk/dpdkAsmOpt.cpp (hunk 4 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkAsmOpt.cpp]: done

File: backends/dpdk/dpdkProgram.cpp [hunk 5]
```diff
@@ -34,13 +33,9 @@ IR::IndexedVector<IR::DpdkStructType> ConvertToDpdkProgram::UpdateHeaderMetadata
     for (auto obj : prog->objects) {
         if (auto s = obj->to<IR::Type_Struct>()) {
             if (s->name.name == structure->local_metadata_type) {
```
Comment (usha1830) [inline backends/dpdk/dpdkProgram.cpp:35]: Why is this removed? Where are the annotations added ?

**Review by usha1830 — APPROVED** [review]

File: backends/dpdk/dpdkAsmOpt.cpp [hunk 6]
```diff
@@ -163,4 +163,41 @@ const IR::IndexedVector<IR::DpdkAsmStatement> *RemoveLabelAfterLabel::removeLabe
...
+    auto usedFieldsCollector = new CollectUsedMetadataField();
+    usedFieldsCollector->setCalledBy(this);
+    p->apply(*usedFieldsCollector);
+    auto used_fields = usedFieldsCollector->getUsedFields();
+
+    IR::IndexedVector<IR::DpdkStructType> usedStruct;
+    bool isMetadataStruct = false;
+    for (auto st : p->structType) {
+        if (!isMetadataStruct) {
+            for (auto anno : st->annotations->annotations) {
```
Comment (psivanup) [inline backends/dpdk/dpdkAsmOpt.cpp:177]: Can st->annotations be null in any case? It would be better if it is checked for nullptr before continue.

**Review by psivanup — CHANGES_REQUESTED** [review]

**Comment by kamleshbhalui** [comment]
It will never be null, all annotations assigned with this initially , new Annotations(Vector<Annotation>());

**Review by psivanup — APPROVED** [review]
